frontend/src/pages/QueryBuilder/QueryBuilder.tsx

```diff
@@ -127,30 +155,61 @@ class QueryBuilder extends React.Component<QueryBuilderProps> {
     const editable = this.isEditable(operation);
 
     return (
-      <OperationForm
-        operation={operation}
-        editable={editable}
-        valid={steps.count() > 0}
-        onUpdateOperation={this.onUpdateOperation}
-        onDuplicateOperation={this.onDuplicateOperation}
-        onSuccess={this.onSaveOperation}
-        processing={this.props.page.get('processing') as boolean}
-        onDeleteOperation={this.onDeleteOperation}
-        onReset={!id ? () => this.props.actions.setActiveOperation() : undefined}
-      >
-        <OperationSteps
-          sources={this.props.sources.get('sources') as List<SourceMap>}
-          isFetchingSources={this.props.sources.get('loading') as boolean}
-          steps={steps}
-          fetchSources={this.props.actions.fetchSources}
-          onSelectSource={this.props.actions.setActiveSource}
-          onAddStep={this.props.actions.updateActiveStep}
-          activeSource={this.props.activeSource}
-          activeStep={activeStep}
-          onClickStep={(step) => this.props.actions.updateActiveStep(step, true)}
+      <>
+        <OperationForm
+          operation={operation}
           editable={editable}
-        />
-      </OperationForm>
+          valid={steps.count() > 0}
+          onUpdateOperation={this.onUpdateOperation}
+          onDuplicateOperation={this.onDuplicateOperation}
+          onSuccess={this.onSaveOperation}
+          onPreview={this.onPreviewOperation}
+          previewing={this.state.loadingPreview}
+          processing={this.props.page.get('processing') as boolean}
+          onDeleteOperation={this.onDeleteOperation}
+          onReset={!id ? () => this.props.actions.setActiveOperation() : undefined}
+        >
+          <OperationSteps
+            sources={this.props.sources.get('sources') as List<SourceMap>}
+            isFetchingSources={this.props.sources.get('loading') as boolean}
+            steps={steps}
+            fetchSources={this.props.actions.fetchSources}
+            onSelectSource={this.props.actions.setActiveSource}
+            onAddStep={this.addStep.bind(this)}
+            activeSource={this.props.activeSource}
+            activeStep={activeStep}
+            onClickStep={(step) => {
+              this.setState((state) => {
+                return { ...state, previewShow: false };
+              });
+              this.props.actions.updateActiveStep(step, true);
+            }}
```

Comment:
As it was with steps, if previewing, the form should be disabled i.e no clicking steps until someone closes the preview ... keeps things consistent. Have updated this view to better handle this in another branch, but no worries ... I'll deal with the conflicts.